core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ast/StatementMatcher.java

```diff
@@ -44,6 +46,73 @@ public StatementMatcher(Variable subject, Variable predicate, Variable object) {
 		this.objectValue = object == null ? null : object.value;
 	}
 
+	public static List<StatementMatcher> reduce(List<StatementMatcher> statementMatchers) {
+		List<StatementMatcher> wildcardMatchers = statementMatchers.stream()
+				.filter(s -> s.subjectIsWildcard() || s.predicateIsWildcard() || s.objectIsWildcard())
+				.collect(Collectors.toList());
+
+		if (wildcardMatchers.isEmpty()) {
+			return statementMatchers;
+		}
+
+		return statementMatchers.stream().filter(s -> {
+			for (StatementMatcher statementMatcher : wildcardMatchers) {
+				if (statementMatcher != s && statementMatcher.covers(s)) {
+					return false;
+				}
+			}
+
+			return true;
+		}).collect(Collectors.toList());
+
+	}
+
+	private boolean covers(StatementMatcher s) {
+		boolean covers = true;
+
+		if (subjectIsWildcard()) {
+			if (s.subjectName != null) {
+				covers = false;
+			}
+		} else {
+			if (!Objects.equals(subjectName, s.subjectName)) {
+				covers = false;
+			}
+			if (!Objects.equals(subjectValue, s.subjectValue)) {
+				covers = false;
+			}
+		}
+
+		if (predicateIsWildcard()) {
+			if (s.predicateName != null) {
+				covers = false;
+			}
+		} else {
+			if (!Objects.equals(predicateName, s.predicateName)) {
+				covers = false;
+			}
+			if (!Objects.equals(predicateValue, s.predicateValue)) {
+				covers = false;
+			}
+		}
+
+		if (objectIsWildcard()) {
+			if (s.objectName != null) {
+				covers = false;
+			}
+		} else {
+			if (!Objects.equals(objectName, s.objectName)) {
+				covers = false;
+			}
+			if (!Objects.equals(objectValue, s.objectValue)) {
+				covers = false;
+			}
+		}
```

Comment:
Hi, Not sure how often this is called but once the `covers` is determined to be false the rest of the checks does not makes sense.

Reply:
Good point!